pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+		glog.Errorf("Error releasing ip %v for service %v: %v", ipString, key, err)
+		return
+	}
+	delete(ic.allocationMap, ipString)
+}
+
+// recordAllocation updates local state with the ingress ip indicated
+// in a service's status and ensures that the ingress ip appears in
+// the service's list of external ips.
+func (ic *IngressIPController) recordAllocation(service *kapi.Service, key, oldIP string) error {
+	// If more than one ingress ip is present, it will be ignored
+	ipString := service.Status.LoadBalancer.Ingress[0].IP
+
+	// Clear a previously allocated ip
+	if oldIP != "" && ipString != oldIP {
+		ic.clearLocalAllocation(key, oldIP)
+	}
+
+	// Make a copy to modify to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(service)
+	if err != nil {
+		return err
+	}
+	sCopy := t.(*kapi.Service)
+
+	ip := net.ParseIP(ipString)
+	if ip == nil {
+		// An invalid ingress ip is assigned to the service.
+		if !ic.syncProcessed {
+			// Fix the problem after processing the initial sync to
+			// avoid delaying cache updates any longer than necessary.
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			return ic.clearPersistedAllocation(sCopy)
+		}
+	}
+
+	ipKey, ok := ic.allocationMap[ipString]
+	if ok && ipKey == key {
+		// Allocation is for this service.
+		if !ic.syncProcessed {
+			// Avoid ensuring an external ip until after the initial sync.
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			// Ensure that the ingress ip is present in the service's spec.
+			return ic.ensureExternalIP(sCopy, ipString)
+		}
+	}
+
+	err = ic.ipAllocator.Allocate(ip)
+	if err != nil {
+		if !ic.syncProcessed {
+			// Fix the problem after processing the initial sync
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			// Clear the allocation so the service can be re-allocated
+			// when the update is seen.
+			return ic.clearPersistedAllocation(sCopy)
+		}
+	}
+	ic.allocationMap[ipString] = key
+	if !ic.syncProcessed {
+		// Avoid ensuring an external ip until after the initial sync.
+		ic.pendingAllocations = append(ic.pendingAllocations, service)
+		return nil
+	} else {
+		// Ensure that the ingress ip is present in the service's spec.
+		return ic.ensureExternalIP(sCopy, ipString)
+	}
+}
+
+// clearPersistedAllocation ensures there is no ingress ip in the
+// service's spec and that the service's status is cleared.
+func (ic *IngressIPController) clearPersistedAllocation(service *kapi.Service) error {
+	// Assume it is safe to modify the service without worrying about changing the local cache
+
+	// An ingress ip is only allowed in ExternalIPs when a
+	// corresponding status exists, so update the spec first to avoid
+	// failing admission control.
+	ingressIP := service.Status.LoadBalancer.Ingress[0].IP
+
+	filteredIPs := service.Spec.ExternalIPs[:0]
```

Comment:
This might be simpler:
```
for i, ip := range service.Spec.ExternalIPs {
       if ip == ingressIP {
            service.Spec.ExternalIPs = append(service.Spec.ExternalIPs[:i], service.Spec.ExternalIPs[i+1:])
           if err := ic.persistSpec(service); err != nil {
                return err
           }
}
```